core/src/main/java/com/linecorp/armeria/common/multipart/SubscriptionArbiter.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (c)  2020 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.linecorp.armeria.common.multipart;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscription;
+
+import com.google.common.math.LongMath;
+
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
+
+/**
+ * Allows changing the subscription, tracking requests and item
+ * production in concurrent source-switching scenarios.
+ * <p>
+ *     {@code this} is the work-in-progress indicator for the
+ *     subscriber-request-produced trampolining.
+ * </p>
+ * <p>
+ *     Please override {@link #request(long)} and perform the n &lt;= 0L
+ *     check in the context of the implementor because the TCK requires
+ *     an onError signal and the arbiter has no contextual knowledge how
+ *     and when to signal it.
+ * </p>
+ */
+class SubscriptionArbiter extends AtomicInteger implements Subscription {
```

Comment:
- We could make all `Atomic*` fields volatile with `Atomic*FieldUpdater`.
- Similarly, we don't need to extend `AtomicInteger`.